src/data/real/liouville.lean

```diff
@@ -28,34 +24,143 @@ A Liouville number is a real number `x` such that for every natural number `n`,
+  -- Actually, `q` is a natural number
   lift q to ℕ using (zero_lt_one.trans q1).le,
+  -- Looks innocuous, but we now have an integer with non-zero absolute value: this is at
+  --least one away from zero.  The gain here is what gets the proof going.
   have ap : 0 < abs (a * ↑q - ↑b * p) := abs_pos.mpr a0,
+  -- Actually, the absolute value of an integer is a natural number
   lift (abs (a * ↑q - ↑b * p)) to ℕ using (abs_nonneg (a * ↑q - ↑b * p)),
+  -- At a1, revert to natural numbers
   rw [← int.coe_nat_mul, ← int.coe_nat_pow, ← int.coe_nat_mul, int.coe_nat_lt] at a1,
+  -- Recall this is by contradiction: we obtained the inequality `b * q ≤ x * q ^ (b + 1)`, so
+  -- we are done.
   exact not_le.mpr a1 (nat.mul_lt_mul_pow_succ (int.coe_nat_pos.mp ap) (int.coe_nat_lt.mp q1)).le,
 end
 
-end is_liouville
+end liouville
+
+open polynomial metric set real ring_hom
+
+/-- This lemma collects the properties needed to prove `exists_pos_real_of_irrational_root`.
+It is stated in more general form than needed: in the intended application, `Z = ℤ`, `N = ℕ`,
+`R = ℝ`, `d a = a ^ f.nat_degree`, `j z a  = z / (a + 1)`, `f ∈ ℤ[x]`, `α` is an irrational
+root of `f`, `ε` is small, `M` is a bound on the Lipschitz constant of `f` near `α`, `n` is
+the degree of the polynomial `f`.
+-/
+lemma exists_one_le_pow_mul_dist {Z N R : Type*} [metric_space R]
+  {d : N → ℝ} {j : Z → N → R} {f : R → R} {α : R} {ε M : ℝ}
+-- denominators are positive
+  (d0 : ∀ (a : N), 1 ≤ d a)
+  (e0 : 0 < ε)
+--function is Lipschitz at α
+  (B : ∀ ⦃y : R⦄, y ∈ closed_ball α ε → dist (f α) (f y) ≤ (dist α y) * M)
+-- clear denominators
+  (L : ∀ ⦃z : Z⦄, ∀ ⦃a : N⦄, j z a ∈ closed_ball α ε → 1 ≤ (d a) * dist (f α) (f (j z a))) :
+  ∃ e : ℝ, 0 < e ∧ ∀ (z : Z), ∀ (a : N), 1 ≤ (d a) * (dist α (j z a) * e) :=
+begin
+  -- A useful inequality to keep at hand
+  have me0 : 0 < max (1 / ε) M := lt_max_iff.mpr (or.inl (one_div_pos.mpr e0)),
+  -- The maximum between `1 / ε` and `M` works
+  refine ⟨max (1 / ε) M, me0, λ z a, _⟩,
+  -- First, let's deal with the easy case in which we are far away from `α`
+  by_cases dm1 : 1 ≤ (dist α (j z a) * max (1 / ε) M),
+  { exact one_le_mul_of_one_le_of_one_le (d0 a) dm1 },
+  { -- `j z a = z / (a + 1)`: we prove that this ratio is close to `α`
+    have : j z a ∈ closed_ball α ε,
+    { refine mem_closed_ball'.mp (le_trans _ ((one_div_le me0 e0).mpr (le_max_left _ _))),
+      exact ((le_div_iff me0).mpr (not_le.mp dm1).le) },
+    -- use the "separation from `1`" (assumption `L`) for numerators,
+    refine (L this).trans _,
+    -- remove a common factor and use the Lipschitz assumption `B`
+    refine mul_le_mul_of_nonneg_left ((B this).trans _) (zero_le_one.trans (d0 a)),
+    exact mul_le_mul_of_nonneg_left (le_max_right _ M) dist_nonneg }
+end
+
+lemma da {R : Type*} {s t : set R} {x : R} (h : s = t) (hx : x ∈ s) :
+  x ∈ t :=
+begin
+  have : s ⊆ t, exact h.subset,
+  exact h.subset hx,
+end
```

Comment:
Remove?

Reply:
Ooops, this should not have stayed there!  Thanks!